### DISCUSSION TITLE
Custom pub/sub events?

### DISCUSSION THREAD

**Discussion by cj** [zauberzeug/nicegui#662] [Q&A] [answered]
I tried looking for something in the docs, but is there a way to publish and subscribe to custom events on the client and server? I saw this https://github.com/zauberzeug/nicegui/issues/644. So, if I wanted to create a simple chat room, I could subscribe to the custom web socket channel and publish messages directly to it. Or should I use something like pusher.com for that?

**Comment by rodja** [comment]
Thanks for reaching out. When implementing your [own custom component](https://github.com/zauberzeug/nicegui/tree/main/examples/custom_vue_component) it's fairly easy to send custom events from the backend (Python):

https://github.com/zauberzeug/nicegui/blob/3278fa93a33e1d2f3053ceba8e03f244f70b3453/examples/custom_vue_component/counter.py#L17

to frontend (JavaScript):

https://github.com/zauberzeug/nicegui/blob/3278fa93a33e1d2f3053ceba8e03f244f70b3453/examples/custom_vue_component/counter.js#L16-L18

and the other way around; from frontend:

https://github.com/zauberzeug/nicegui/blob/3278fa93a33e1d2f3053ceba8e03f244f70b3453/examples/custom_vue_component/counter.js#L14

to backend:

https://github.com/zauberzeug/nicegui/blob/3278fa93a33e1d2f3053ceba8e03f244f70b3453/examples/custom_vue_component/counter.py#L14

**Reply by cj** [reply]
@rodja thank you for the response. I'm sorry if I wasn't clear in my question. I have no issue creating custom components; I wanted to be able to create custom events in the underlying NiceGUI socket.io implementation that is used to update the UI in real-time for the user and avoid using https://pusher.com or implementing my own socket.io transit. Does that make sense?

**Reply by cj** [reply]
For example, this video explains creating a simple chat app using socket.io https://www.youtube.com/watch?v=BSGGAzsgFvk. But instead of initializing socket.io again and making another connection from the client to the server, I could just subscribe and publish using NiceGUI's established connection.

**Reply by rodja** [reply]
Yes. That is exactly what I was trying to say above. The custom event example is using the same socket.io connection as the rest of NiceGUI. Have a look at https://github.com/zauberzeug/nicegui/blob/main/nicegui/elements/log.py which implements a log view. Very similar you could implement a chat element which receives messages and renders them. But that would be overkill in a scenario like a chat application. You could simply use `ui.label` to add new messages and `ui.input` for the text entry.

**Reply by cj** [reply]
Ah, I 100% understand now; thank you for the explanation and code references! 😄

**Comment by rodja** [comment, marked as answer]
@cj I just added a [chat app example](https://github.com/zauberzeug/nicegui/blob/main/examples/chat_app/main.py) which does not need any special pub/sub. Just NiceGUI standard functionality.

**Reply by cj** [reply]
@rodja amazing! Thank you for that. I'm blown away by how nice (pun intended), NiceGUI is to work with. Once #658 is merged, it will make it much more lightweight for smaller apps and mobile.

**Comment by cj** [comment]
@rodja coming back to this subject real quick. Let us say you have it hooked up to a database and wanted to avoid polling it for performance reasons. Is there a NiceGUI way to only submit an event to certain people?

**Reply by rodja** [reply]
You will need a database/storage which provides an event mechanism. Like Redis for example. Then you could register your [update_messages function](https://github.com/zauberzeug/nicegui/blob/5f39e4a42c9cbebf14d87912fb0d3c40eb73a51d/examples/chat_app/main.py#L26) with the specific db event. NiceGUI will automatically find the right context to only send the ui commands to the client which is currently viewing this page.

**Reply by cj** [reply]
Ah ok - that's what I thought, and why I mentioned pusher.com/socket.io before (https://pusher.com/docs/channels/using_channels/events/). It would have been nice to re-use the same socket.io connection already established.

**Reply by rodja** [reply]
A socket.io connection is always between two parties. So either between the client and the backend or between the backend and a database. Surely you could also create a third connection directly between database and frontend. But with NiceGUI I would advise to play everything through Python (eg. backend). That will give you the most flexibility and does not require any custom components.

**Reply by cj** [reply]
Correct, but you can set socket.io up so you can join "rooms" i.e., subscribe to `chat/:uuid,` lots of people can all be connected to the same room, and you can then emit events on it that everyone in that room will get. https://socket.io/docs/v4/rooms/. So in this case, where you have the timer checking for update_messages https://github.com/zauberzeug/nicegui/blob/5f39e4a42c9cbebf14d87912fb0d3c40eb73a51d/examples/chat_app/main.py#L33. You could subscribe to that socket.io room, then have anyone in that room trigger say a message sent event, which will then call update_messages and then fetch the all the messages from the DB. This can all be done via the socket.io connection already established by NiceGUI, if the socket.io implementation was exposed.

**Reply by rodja** [reply]
Oh, thats a great idea @cj. I updated the [chat app example](https://github.com/zauberzeug/nicegui/blob/main/examples/chat_app/main.py) accordingly. As you can see, there is still no need for additional socket communications. I just keep the ui element of each chat-participant in a list. Whenever a new message arrives, all chat views are updated. Thereby we can get rid of the timer.

**Comment by cj** [comment]
@rodja awesome - thank you for the update!

I am creating a single-page application https://github.com/zauberzeug/nicegui/tree/main/examples/single_page_app/. So I could use this same concept from your chat example for global app updates to update certain page sections like the on_connected https://github.com/zauberzeug/nicegui/blob/cff68385ec643db7ec85da9425655cea9f8f2222/nicegui/globals.py#L53. i.e. when a user logs in, I then update the header to show they are logged in, or to make global app toast notifications?

I am also getting this type warning with the example https://github.com/zauberzeug/nicegui/blob/main/examples/chat_app/main.py#L31-L32:
<img width="758" alt="image" src="https://user-images.githubusercontent.com/1819/229691561-d602342d-1202-423a-9f13-c202755702a1.png">

**Reply by rodja** [reply]
> . i.e. when a user logs in, I then update the header to show they are logged in, or to make global app toast notifications?

Is that a multi-user application where you want to notify others who is online? Then the approach you described sounds right.

I think it would be good to open a new discussion for the type warning. It looks strange.

**Reply by cj** [reply]
@rodja, can you open the SocketManager so we can pass in **kwargs https://github.com/pyropy/fastapi-socketio/pull/30? Maybe add an option to `run`? https://github.com/zauberzeug/nicegui/blob/6cf53c042152a7c4c3fce3238a34272a7f5e110f/nicegui/nicegui.py#L27

This would allow us to use a Redis adapter, as we need to be able to emit socketio events from workers/jobs.

**Reply by rodja** [reply]
Sounds good. Would you like to create a pull request?
